### PR TITLE
Add check_mode to get_url 

### LINKED ISSUE #20531: get_url should check for URL existence in check mode
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
get_url plugin/library

##### ANSIBLE VERSION

```
ansible 2.2.0.0
```

##### OS / ENVIRONMENT

N/A

##### SUMMARY

Check mode for get_url should make sure the URL to fetch exists

##### STEPS TO REPRODUCE

```yaml

```
Ran get_url using variables to construct a URL that did not exist for some hosts

##### EXPECTED RESULTS

Expected check mode to fail on nonexistent URL

##### ACTUAL RESULTS

Check mode is skipped

```

```

### PR DESCRIPTION
**Pull request by SesquipedalianDefenestrator**
Add check_mode to get_url that does a HEAD request to make sure the URL exists, but doesn't write the real file

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

get_url
##### ANSIBLE VERSION

```
ansible-2.2.0.0
```

##### SUMMARY

Fixes #20531 
To make sure the URL exists in check mode, do a HEAD request and write to the temp location, then clean it up and exit without copying to the real destination.

```

```

### CODE REVIEW

File: lib/ansible/modules/network/basics/get_url.py
Comment (robinro) [inline lib/ansible/modules/network/basics/get_url.py]: Add a trailing comma to reduce the diff when arguments are added in the future.

**Review by robinro — CHANGES_REQUESTED** [review]
Overall the change looks good to me.

In general it assumes `HEAD` is supported by the server, which must not be the case. I see the benefit greater than this issue.

I'd suggest to
* add tests for the new behavior
  * get_url to existing URL in normal and check_mode
  * get_url to non-existing URL in normal and check_mode
* mention the behavior in the module documentation; an alternative expectation could be that the URL is fetched and compared against the destination/checksum/...

**Comment by SesquipedalianDefenestrator** [comment]
Good point on the destination/checksum comparison.  Downloading the whole file could potentially add surprising load to the source as well as using resources and space on the destination, increasing the potential to break things with --check.  For our particular use, HEAD is the better tradeoff, but I'm open to being persuaded that it's the wrong behavior in the general case.

File: test/integration/targets/get_url/tasks/main.yml
Comment (SesquipedalianDefenestrator) [inline test/integration/targets/get_url/tasks/main.yml]: Let me know if there's a "guaranteed to 404" URL I should use instead.  Using a test domain guarantees this test won't magically break via 3rd-party action, but the NXDOMAIN can be a bit slow.

**Comment by SesquipedalianDefenestrator** [comment]
Not sure what to do about the failed tests.  Looks like it's a failure to contact AWS from the build infrastructure or something?
2017-01-25 01:25:34 Run command: curl -s -S -i -X GET -H 'Expect: ' https://14blg63h2i.execute-api.us-east-1.amazonaws.com/prod/jobs/a8cab045-6bec-4605-a9e4-fb9456bb26df
2017-01-25 01:25:35 Retrieved starting freebsd/10.3-STABLE instance a8cab045-6bec-4605-a9e4-fb9456bb26df.
2017-01-25 01:25:45 ERROR: Timeout waiting for freebsd/10.3-STABLE instance a8cab045-6bec-4605-a9e4-fb9456bb26df.

File: test/integration/targets/get_url/tasks/main.yml
Comment (robinro) [reply test/integration/targets/get_url/tasks/main.yml]: @sivel You're the httptester expert. Any suggestion for this test?

@SesquipedalianDefenestrator We should keep external dependencies minimal, so there should be a httptester solution or something local.

Probably http://{{httpbin_host}/DOESNOTEXIST is a good-enough URL.

**Comment by SesquipedalianDefenestrator** [comment]
I guess the test failure was something in the test infrastructure.  Moving around spacing and changing the URL seems to have triggered a re-run, which passed.

**Comment by SesquipedalianDefenestrator** [comment]
Anything else that needs changing/updating on this?  Wanted to make sure it's not waiting on something from me, since it's marked "needs_revision" and affects_2.3 (which is released.. )

**Review by robinro — APPROVED** [review]
LGTM
Thanks @SesquipedalianDefenestrator

**Comment by robinro** [comment]
@jpmens Please review and
shipit

**Comment by jpmens** [comment]
shipit

**Comment by gundalow** [comment]
Hi,
A number of files have been git moved from `modules/network` to `modules/net_tools`, which has effected this PR. Please rebase this PR to pull in that change.

**Comment by gundalow** [comment]
If someone pokes me on IRC once this has been rebased I'll merge

**Comment by SesquipedalianDefenestrator** [comment]
Failure is "2017-04-27 20:35:24 ERROR: Timeout waiting for freebsd/10.3-STABLE instance a53ce7df-a649-434b-b036-d9d7302b86ae."  
All other platforms/tests succeeded, so I'm guessing ephemeral failure with the test infra.
Can someone with write access re-run the tests?  Or I can just submit gratuitous whitespace changes to trigger a re-run.

**Comment by gundalow** [comment]
Merged, thanks for the review.

I made one minor change before merging to make it clear that the `--check` feature is in Ansible 2.4 onwards.

As this is a feature (rather than a bug) this isn't going to be backported.

**Pull request merged**
